core/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java

```diff
@@ -207,4 +263,132 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
     public String getName() {
         return NAME;
     }
+
+    @Override
+    protected Query doToQuery(QueryParseContext parseContext) throws IOException {
+        List<Object> terms = new ArrayList<>();
+        if (isTermsLookupQuery()) {
+            handleTermsLookup(new TermsLookup(lookupIndex, lookupType, lookupId, lookupRouting, lookupPath, parseContext), parseContext.client(), terms);
+        } else {
+            terms.addAll(values);
+        }
+        if (terms.isEmpty()) {
+            return Queries.newMatchNoDocsQuery();
+        }
+        return handleTermsQuery(terms, fieldName, parseContext, isFilter);
+    }
+
+    static void handleTermsLookup(TermsLookup lookup, Client client, List<Object> terms) {
+        GetRequest getRequest = new GetRequest(lookup.getIndex(), lookup.getType(), lookup.getId()).preference("_local").routing(lookup.getRouting());
+        getRequest.copyContextAndHeadersFrom(SearchContext.current());
+        final GetResponse getResponse = client.get(getRequest).actionGet();
+        if (getResponse.isExists()) {
+            List<Object> extractedValues = XContentMapValues.extractRawValues(lookup.getPath(), getResponse.getSourceAsMap());
+            terms.addAll(extractedValues);
+        }
+    }
+
+    static Query handleTermsQuery(List<Object> terms, String fieldName, QueryParseContext parseContext, boolean isFilter) {
+        MappedFieldType fieldType = parseContext.fieldMapper(fieldName);
+        String indexFieldName;
+        if (fieldType != null) {
+            indexFieldName = fieldType.names().indexName();
+        } else {
+            indexFieldName = fieldName;
+        }
+
+        Query query;
+        if (isFilter) {  // NO COMMIT: not exactly sure how to handle this case, right now using a filter flag 
```

Comment:
I think you should use QueryShardContext#isFilter but that is something that @cbuescher is working on, he should be able to give you some more details on that